### DISCUSSION TITLE
Is there a way to run an MSBuild target before the first project or after the last project?

### DISCUSSION THREAD

**Discussion by adcloeshelle** [dotnet/msbuild#8074] [Q&A] [answered]
We have a use case where we want to run some pre-build or post-build logic for the entire solution. Is there a way to do this via an MSBuild Task?

**Comment by rainersigwald** [comment, marked as answer]
Yes: https://learn.microsoft.com/visualstudio/msbuild/customize-your-build#customize-the-solution-build. However, it may not do what you want, since it doesn't apply in Visual Studio.

You may instead want to use a `Directory.Build.targets` to inject a target into every project, and inside that target use the `MSBuild` task to build a project that contains your pre-build logic.

**Reply by adcloeshelle** [reply]
Thank you for the quick response! TIL about those solution props/targets files!

For the second option - how does that work? If the target is in every project, wouldn't that mean we'd run the pre-build step for every project? Or the idea is MSBuild is smart enough to know the project is already built so skips building it multiple times?

**Reply by GangWang01** [reply]
> For the second option - how does that work? If the target is in every project, wouldn't that mean we'd run the pre-build step for every project? Or the idea is MSBuild is smart enough to know the project is already built so skips building it multiple times?

[Customize the build for one or more projects](https://learn.microsoft.com/en-us/visualstudio/msbuild/customize-by-directory?view=vs-2022) specially explains. To control whether to run the target in `Directory.Build.targets` for a project, you can add a condition to the target referencing the property set from the project, e.g. `<Target Name="PreBuild" Condition="'$(EnablePreBuild)' == 'true'">`.

**Reply by Nirmal4G** [reply]
Updated Link: https://learn.microsoft.com/visualstudio/msbuild/customize-solution-build
